compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/HashFunction.java

```diff
@@ -0,0 +1,165 @@
+        return Collections.unmodifiableList(instances);
+    }
+
+    private static List<HashFunction> instances = new ArrayList<>();
+
+    private static int[] mersennePrimes = {3, 7, 31, 127, 8191, 131071, 524287, 2147483647};
+
+    static {
+      //@formatter:off
+
+        add("val", 0,
+            (val, min) -> val,
+            gen -> (val, min) -> val);
+
+        add("val - min", 1,
+            (val, min) -> val - min,
+            gen -> (val, min) -> gen.emitSub(val, min, false));
+
+        add("val >> min", 1,
+            (val, min) -> val >> min,
+            gen -> (val, min) -> gen.emitShr(val, min));
+
+        add("val >> (val & min)", 2,
+            (val, min) -> val >> (val & min),
+            gen -> (val, min) -> gen.emitShr(val, gen.emitAnd(val, min)));
+
+        add("(val >> min) ^ val", 2,
+            (val, min) -> (val >> min) ^ val,
+            gen -> (val, min) -> gen.emitXor(gen.emitShr(val, min), val));
+
+        add("(val >> min) * val", 3,
+            (val, min) -> (val >> min) * val,
+            gen -> (val, min) -> gen.emitMul(gen.emitShr(val, min), val, false));
+
+        addWithPrimes("(val * prime) >> min", 3,
+                      prime -> (val, min) -> (val * prime) >> min,
+                      (gen, prime) -> (val, min) -> gen.emitShr(gen.emitMul(val, prime, false), min));
+
+        addWithPrimes("rotateRight(val, prime)", 3,
+                      prime -> (val, min) -> Long.rotateRight(val, prime),
+                      (gen, prime) -> (val, min) -> gen.emitRor(val, prime));
+
+        addWithPrimes("rotateRight(val, prime) + val", 4,
+                      prime -> (val, min) -> Long.rotateRight(val, prime) + val,
+                      (gen, prime) -> (val, min) -> gen.emitAdd(gen.emitRor(val, prime), val, false));
+
+        addWithPrimes("rotateRight(val, prime) ^ val", 4,
+                      prime -> (val, min) -> Long.rotateRight(val, prime) ^ val,
+                      (gen, prime) -> (val, min) -> gen.emitXor(gen.emitRor(val, prime), val));
+      //@formatter:on
+    }
+
+    private static void add(String toString, int effort, BiFunction<Long, Long, Long> f, Function<ArithmeticLIRGenerator, BiFunction<Value, Value, Value>> gen) {
+        instances.add(new HashFunction() {
+
+            @Override
+            public int apply(long value, long min) {
+                return f.apply(value, min).intValue();
+            }
+
+            @Override
+            public int effort() {
+                return effort;
+            }
+
+            @Override
+            public String toString() {
+                return toString;
+            }
+
+            @Override
+            public Value gen(Value val, Value min, ArithmeticLIRGenerator t) {
+                return gen.apply(t).apply(toDWORD(val, t), toDWORD(min, t));
+            }
+
+            private Value toDWORD(Value v, ArithmeticLIRGenerator t) {
+                if (v.getPlatformKind() != AMD64Kind.DWORD) {
```

Comment:
This use of AMD64Kind in a shared file should be fixed in the next PR.  I would in general expect the values to come into this function with the right sizing of either int or long with the caller ensuring that it was at least those types already.

Reply:
Why wait for the next PR? Best to fix this here.

Reply:
Let me fix it. Should I create a new PR with the fix or are you going to reopen this one?

Reply:
Create a new PR.